sycl/plugins/unified_runtime/ur/adapters/cuda/enqueue.cpp

```diff
@@ -284,9 +283,32 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunch(
   UR_ASSERT(workDim > 0, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
   UR_ASSERT(workDim < 4, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
 
+  std::vector<ur_event_handle_t> DepEvents(
+      phEventWaitList, phEventWaitList + numEventsInWaitList);
+
+  // phEventWaitList only contains events that are handed to UR by the SYCL
+  // runtime. However since UR handles memory dependencies within a context
+  // we may need to add more events to our dependent events list
+  for (auto &MemArg : hKernel->Args.MemObjArgs) {
+    MemArg.Mem->MemoryMigrationMutex.lock();
```

Comment:
Are you sure the check for `isBuffer` requires locking, if not you could fold it to after the `if` and save the construction of the lock for non buffer arguments.